cpp/src/arrow/filesystem/path_util.cc

```diff
@@ -287,6 +288,38 @@ bool IsLikelyUri(util::string_view v) {
   return ::arrow::internal::IsValidUriScheme(v.substr(0, pos));
 }
 
+struct Globber::Impl {
+  std::regex pattern_;
+
+  explicit Impl(std::string p) : pattern_(std::regex(std::move(transform(p)))) {}
+
+  std::string transform(std::string p) {
```

Comment:
Call this e.g. `PatternToRegex`? Also, can take a `const std::string&` to avoid a pointless copy.